File: android/app/src/main/java/com/rishikesh/edgechat/data/ConversationStore.kt
package com.rishikesh.edgechat.data

import android.content.Context
import com.rishikesh.edgechat.model.Attachment
import com.rishikesh.edgechat.model.Conversation
import com.rishikesh.edgechat.model.Iso
import com.rishikesh.edgechat.model.Role
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.serialization.json.Json
import java.io.File
import java.util.concurrent.Executors

/** JSON-on-disk conversation persistence: files/Conversations/<id>.json (+ <id>/ for attachments). Main-thread API. */
class ConversationStore(context: Context) {
    val root: File = File(context.filesDir, "Conversations").apply { mkdirs() }
    private val _conversations = MutableStateFlow<List<Conversation>>(emptyList())
    val conversations: StateFlow<List<Conversation>> = _conversations
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true; prettyPrint = false }
    /** Serial so writes land in the order they were requested. */
    private val saveExecutor = Executors.newSingleThreadExecutor { r -> Thread(r, "edgechat-save") }

    fun loadAll() {
        val files = root.listFiles { f -> f.extension == "json" } ?: emptyArray()
        val list = files.mapNotNull { f -> runCatching { json.decodeFromString<Conversation>(f.readText()) }.getOrNull() }
            .map { c ->
                // A reply the app was killed in the middle of leaves an empty bubble; drop it.
                c.copy(messages = c.messages.filterNot { it.role == Role.ASSISTANT && it.content.isEmpty() && it.error == null && it.stats == null })
            }
            .sortedByDescending { Iso.parse(it.updatedAt) }
        _conversations.value = list
    }

    fun conversation(id: String): Conversation? = _conversations.value.firstOrNull { it.id == id }

    fun create(): Conversation {
        val c = Conversation()
        _conversations.update { listOf(c) + it }
        save(c)
        return c
    }

    /** `persist = false` updates memory only (streaming token flushes); call again with `persist = true` to write. */
    fun update(conversation: Conversation, touch: Boolean = true, persist: Boolean = true) {
        val c = if (touch) conversation.copy(updatedAt = Iso.now()) else conversation
        _conversations.update { list ->
            val i = list.indexOfFirst { it.id == c.id }
            val next = if (i >= 0) list.toMutableList().also { it[i] = c } else (listOf(c) + list).toMutableList()
            if (touch) next.sortedByDescending { Iso.parse(it.updatedAt) } else next
        }
        if (persist) save(c)
    }

    fun modify(id: String, touch: Boolean = true, persist: Boolean = true, body: (Conversation) -> Conversation) {
        val c = conversation(id) ?: return
        update(body(c), touch, persist)
    }

    fun delete(id: String) {
        _conversations.update { list -> list.filterNot { it.id == id } }
        fileFor(id).delete()
        attachmentsDirectory(id).deleteRecursively()
    }

    fun attachmentsDirectory(id: String): File = File(root, id).apply { mkdirs() }
    fun memoryIndexFile(id: String): File = File(attachmentsDirectory(id), "memory-index.json")
    fun attachmentFile(conversationID: String, attachment: Attachment): File = File(attachmentsDirectory(conversationID), attachment.storedFileName)

    private fun fileFor(id: String) = File(root, "$id.json")

    private fun save(c: Conversation) {
        val target = fileFor(c.id)
        saveExecutor.execute {
            runCatching {
                val tmp = File(target.parentFile, "${target.name}.tmp")
                tmp.writeText(json.encodeToString(Conversation.serializer(), c))
                if (!tmp.renameTo(target)) { target.delete(); tmp.renameTo(target) }
            }
        }
    }
}

File: App/EdgeChat/Services/ConversationStore.swift
import Foundation
import EdgeChatCore

/// JSON-on-disk conversation persistence: Documents/Conversations/<id>.json (+ <id>/ for attachments).
@MainActor @Observable
final class ConversationStore {
    private(set) var conversations: [Conversation] = []
    let root: URL

    init() {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        root = docs.appendingPathComponent("Conversations", isDirectory: true)
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    }

    func loadAll() {
        let files = (try? FileManager.default.contentsOfDirectory(at: root, includingPropertiesForKeys: nil)) ?? []
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        conversations = files.filter { $0.pathExtension == "json" }
            .compactMap { try? decoder.decode(Conversation.self, from: Data(contentsOf: $0)) }
            .map { c in
                // A reply the app was killed in the middle of leaves an empty bubble; drop it.
                var c = c
                c.messages.removeAll { $0.role == .assistant && $0.content.isEmpty && $0.error == nil && $0.stats == nil }
                return c
            }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    func conversation(_ id: UUID) -> Conversation? { conversations.first { $0.id == id } }

    @discardableResult
    func create() -> Conversation {
        let c = Conversation()
        conversations.insert(c, at: 0)
        save(c)
        return c
    }

    /// `persist: false` updates memory only (used for streaming token flushes); call again with `persist: true` to write.
    func update(_ conversation: Conversation, touch: Bool = true, persist: Bool = true) {
        var c = conversation
        if touch { c.updatedAt = Date() }
        if let i = conversations.firstIndex(where: { $0.id == c.id }) {
            conversations[i] = c
        } else {
            conversations.insert(c, at: 0)
        }
        if touch { conversations.sort { $0.updatedAt > $1.updatedAt } }
        if persist { save(c) }
    }

    func modify(_ id: UUID, touch: Bool = true, persist: Bool = true, _ body: (inout Conversation) -> Void) {
        guard var c = conversation(id) else { return }
        body(&c)
        update(c, touch: touch, persist: persist)
    }

    func delete(_ id: UUID) {
        conversations.removeAll { $0.id == id }
        try? FileManager.default.removeItem(at: fileURL(id))
        try? FileManager.default.removeItem(at: attachmentsDirectory(id))
    }

    func attachmentsDirectory(_ id: UUID) -> URL {
        let dir = root.appendingPathComponent(id.uuidString, isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    func memoryIndexURL(_ id: UUID) -> URL {
        attachmentsDirectory(id).appendingPathComponent("memory-index.json")
    }

    func attachmentURL(conversationID: UUID, attachment: Attachment) -> URL {
        attachmentsDirectory(conversationID).appendingPathComponent(attachment.storedFileName)
    }

    private func fileURL(_ id: UUID) -> URL { root.appendingPathComponent("\(id.uuidString).json") }

    /// Serial so writes land in the order they were requested (a concurrent queue let a stale flush overwrite the final stats).
    private static let saveQueue = DispatchQueue(label: "edgechat.conversation-save", qos: .utility)

    private func save(_ c: Conversation) {
        let url = fileURL(c.id)
        Self.saveQueue.async {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            encoder.outputFormatting = [.sortedKeys]
            if let data = try? encoder.encode(c) { try? data.write(to: url, options: .atomic) }
        }
    }
}
